src/game/input/input.rs

```diff
@@ -0,0 +1,44 @@
+use ggez::event::Keycode;
+use queues::IsQueue;
+use std::collections::HashSet;
+
+use super::ring_buffer::RingBuffer;
+
+const HISTORY_CAPACITY: usize = 10;
+
+pub struct Input {
+    current_input: HashSet<Keycode>,
+    input_history: RingBuffer<HashSet<Keycode>>,
+}
+
+impl Input {
+    pub fn new() -> Input {
+        Input {
+            current_input: HashSet::new(),
+            input_history: RingBuffer::new(HISTORY_CAPACITY),
+        }
+    }
+
+    pub fn add_input(&mut self, keycode: Keycode) {
+        self.current_input.insert(keycode); 
+    }
+
+    pub fn remove_input(&mut self, keycode: Keycode) {
+        self.current_input.remove(&keycode);
+    }
+
+    pub fn get_input_history(&mut self) -> Vec<HashSet<Keycode>> {
```

Comment:
So, it looks like doing this requires a lifetime to be specified for the return value, if I'm not mistaken. I'm still shaky on how those get implemented exactly, so I'll fix the first comment now while I do some research.

Reply:
I don't think this is correct.  I meant switch from cloning the Vec to just returning a reference to the Vec that is already in self. Since the function only takes in self, and returns a reference, the lifetime parameter should be elided. The only lifetime that makes sense is the return's lifetime must be a subset of self's lifetime.

Reply:
Oh, I see, I think I misread the original comment. Pushing new changes.